### PR TITLE
Add perceiver IO

### LINKED ISSUE #72: [Paper] Perceiver IO: A General Architecture for Structured Inputs & Outputs
- Link: https://arxiv.org/abs/2107.14795

### PR DESCRIPTION
**Pull request by alvanli**
close #72

### CODE REVIEW

File: vformer/attention/cross.py [hunk 1]
```diff
@@ -73,3 +72,57 @@ def forward(self, cls, patches):
...
+        q = self.to_q(x)
+        k = self.to_k(context)
+        v = self.to_v(context)
+        k = rearrange(k, "b n (h d) -> b h n d", h=h)
+        q = rearrange(q, "b n (h d) -> b h n d", h=h)
+        v = rearrange(v, "b n (h d) -> b h n d", h=h)
+
+        attention = torch.matmul(q, k.transpose(-1, -2)) * self.scale
+
+        if mask is not None:
```
Comment (alvanli) [inline vformer/attention/cross.py:118]: I am not sure if we need mask, I think mask is mainly for NLP models?

File: vformer/models/classification/perceiver_io.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import torch.nn as nn
+
+from ...common import BaseClassificationModel
+from ...decoder import PerceiverIODecoder
+from ...encoder import PerceiverIOEncoder
+from ...utils import MODEL_REGISTRY
+
+
+@MODEL_REGISTRY.register()
+class PerceiverIO(nn.Module):
```
Comment (alvanli) [inline vformer/models/classification/perceiver_io.py:11]: I didn't use BaseClassificationModel because img_size and patch_size are not that relevant in this model (model takes in image as a flattened array of pixels

File: vformer/attention/cross.py (hunk 1 above)
Comment (NeelayS) [reply vformer/attention/cross.py:118]: I think it would be helpful to keep the mask option. Some training strategies for vision transformers do patch level supervision in which having the ability to mask regions may be helpful. For eg, [this paper](https://arxiv.org/abs/2111.06377).

File: vformer/models/classification/perceiver_io.py (hunk 2 above)
Comment (NeelayS) [reply vformer/models/classification/perceiver_io.py:10]: I see; that's alright.

File: vformer/attention/cross.py [hunk 3]
```diff
@@ -73,3 +72,57 @@ def forward(self, cls, patches):
...
+    def __init__(self, query_dim, context_dim, num_heads=8, head_dim=64):
+        super().__init__()
+
+        inner_dim = num_heads * head_dim
+        self.num_heads = num_heads
+        self.scale = head_dim ** -0.5
+        self.to_q = nn.Linear(query_dim, inner_dim)
+        self.to_k = nn.Linear(context_dim, inner_dim)
+        self.to_v = nn.Linear(context_dim, inner_dim)
+        self.to_out = _Projection(inner_dim, query_dim)
```
Comment (abhi-glitchhg) [inline vformer/attention/cross.py:104]: Minor nit:
I dont think we need to have another class for this, we can do the same thing with if else statement-
`self.to_out = nn.Linear(inner_dim, query_dim) if inner_dim == query_dim else nn.Identity()`

File: vformer/decoder/perceiver_io.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                queries_dim,
+                latent_dim,
+                num_heads=num_cross_heads,
+                head_dim=cross_head_dim,
+            ),
+            context_dim=latent_dim,
+        )
+        self.decoder_ff = (
+            PreNorm(queries_dim, FeedForward(queries_dim)) if decoder_ff else None
+        )
```
Comment (abhi-glitchhg) [inline vformer/decoder/perceiver_io.py:60]:
```suggestion
        self.decoder_ff = (
            PreNorm(queries_dim, FeedForward(queries_dim)) if decoder_ff else nn.Identity()
        )
```

File: vformer/functional/norm.py [hunk 5]
```diff
@@ -11,11 +11,17 @@ class PreNorm(nn.Module):
         Attention class
     """
 
-    def __init__(self, dim, fn):
+    def __init__(self, dim, fn, context_dim=None):
```
Comment (abhi-glitchhg) [inline vformer/functional/norm.py:14]: Could you please add the description of `context_dim` in the doc string? 
Thanks.

File: vformer/attention/cross.py (hunk 3 above)
Comment (alvanli) [reply vformer/attention/cross.py]: fixed with [27cb828](https://github.com/SforAiDl/vformer/pull/79/commits/27cb82886fbb715b0aa7ef7e6ce5eaa9577593d6)

File: vformer/functional/norm.py (hunk 5 above)
Comment (alvanli) [reply vformer/functional/norm.py:16]: fixed with [8a2fdcb](https://github.com/SforAiDl/vformer/pull/79/commits/8a2fdcbecc7e094f5a46edf8acb8fe217ea4d9aa)

**Comment by NeelayS** [comment]
Hey @alvanli, everything about this PR looks good. If you could resolve the confilcts, we can then go ahead and merge this.

**Review by NeelayS — APPROVED** [review]
